src/main/java/org/peergos/Server.java

```diff
@@ -13,106 +15,209 @@
+        ProviderStore providers = new RamProviderStore();
+        Kademlia dht = new Kademlia(new KademliaEngine(ourPeerId, providers, records), false);
+        CircuitStopProtocol.Binding stop = new CircuitStopProtocol.Binding();
+        CircuitHopProtocol.RelayManager relayManager = CircuitHopProtocol.RelayManager.limitTo(builder.getPrivateKey(), ourPeerId, 5);
+        builder = builder.addProtocols(List.of(
+                new Ping(),
+                new AutonatProtocol.Binding(),
+                new CircuitHopProtocol.Binding(relayManager, stop),
+                new Bitswap(new BitswapEngine(blocks)),
+                dht));
+
+        Host node = builder.build();
+        node.start().join();
+        System.out.println("Node started and listening on " + node.listenAddresses());
+
+        //            Multiaddr bootstrapNode = Multiaddr.fromString("/dnsaddr/bootstrap.libp2p.io/p2p/QmcZf59bWwK5XFi76CZX8cbJ4BhTzzA3gU1ZjYZcYW3dwt");
+        //            KademliaController bootstrap = builder.getWanDht().get().dial(node, bootstrapNode).getController().join();
+
+        APIService localAPI = new APIService();
+        MultiAddress apiAddress = new MultiAddress(config.getProperty("Addresses", "API"));
+        InetSocketAddress localAPIAddress = new InetSocketAddress(apiAddress.getHost(), apiAddress.getPort());
+
+        int maxConnectionQueue = 500;
+        int handlerThreads = 50;
+        HttpServer apiServer = localAPI.initAndStart(localAPIAddress, node, blocks, maxConnectionQueue, handlerThreads);
+
+        Thread shutdownHook = new Thread(() -> {
+            System.out.println("Stopping server...");
+            try {
+                node.stop().get();
+                apiServer.stop(3); //wait max 3 seconds
+                records.close();
+            } catch (Exception ex) {
+                ex.printStackTrace();
+            }
+        });
+        Runtime.getRuntime().addShutdownHook(shutdownHook);
     }
-    private Config readConfig() throws IOException {
-        String home = System.getenv("HOME");
-        Path configFilePath = Path.of(home, ".ipfs", "config");
+    private Config readConfig(Path configPath) throws IOException {
+        Path configFilePath = configPath.resolve("config");
         File configFile = configFilePath.toFile();
         if (! configFile.exists()) {
-            throw new IllegalStateException("Unable to find ./ipfs/config file");
+            System.out.println("Unable to find ./ipfs/config file. Creating default config");
+            String contents = defaultConfig();
+            Files.write(configFilePath, contents.getBytes(), StandardOpenOption.CREATE);
+            return new Config((HashMap) JSONParser.parse(contents));
         }
         String contents = Files.readString(configFilePath);
         return new Config((HashMap) JSONParser.parse(contents));
     }
+    private String defaultConfig() {
```

Comment:
Cleaner to return the map here which can be easily added to I think, and just serialise it in time for writing, and avoid the serde loop.

Reply:
done